srml/system/src/lib.rs

```diff
@@ -755,19 +766,40 @@ impl<T: Trait> Module<T> {
 	}
 }
 
-/// Weight limit check and increment.
+/// resource limit check.
 #[derive(Encode, Decode, Clone, Eq, PartialEq)]
 pub struct CheckWeight<T: Trait + Send + Sync>(PhantomData<T>);
 
 impl<T: Trait + Send + Sync> CheckWeight<T> {
-	fn internal_check_weight(weight: Weight) -> Result<(), DispatchError> {
+	fn internal_check_weight(info: TransactionInfo) -> Result<Weight, DispatchError> {
 		let current_weight = Module::<T>::all_extrinsics_weight();
-		let next_weight = current_weight.saturating_add(weight);
-		if next_weight > T::MaximumBlockWeight::get() {
-			return Err(DispatchError::Payment)
+		let added_weight = info.weight;
+		let next_weight = current_weight.saturating_add(added_weight);
+		let limit = match info.class {
+			DispatchClass::Operational => T::MaximumBlockWeight::get(),
```

Comment:
Should there be any limit at all for operational transactions?

Reply:
There is: the idea is that there is a maximum block weight limit. Normal transaction are only allowed to fill a quarter of that (or any ratio). The operational ones can fill the whole limit. In essence, the limit for operational ones is now 4 times more.

Reply:
@AlistairStewart is this right? ^^^

Reply:
I think @lsaether @keorn must verify this. But based on my discussion with them it is certainly correct that:

> If a block has X weight units available, in normal conditions, we only allow 1/4 to be filled.

What's new is that:
  - What we added as operational txs that can spike to the maximum. This was concluded based on our discussion @gavofyork and not in the specs. But I think it is a reasonable solution for the problem of _always allowing important reports to get in_.
  - We should also allow tipped transactions to go beyond the limit (specs clearly say that). But since we don't have realistic weight numbers yet it is hard to know exactly how. Although we can also simply add a fixed number of extra weight units to the threshold of tipped transactions and let them compete for it.

Reply:
This wasn't my understanding at all. We were thinking of havin reserved space or even no limit for important transactions. But the 1/4 was the target fullness that we adjusted transaction fees for, nit a hard limit. Tips just prioitize transactions in case the maximum has been reached. But maybe @keorn or @AlfonsoCev have a later design.